[PrimTorch] Add reference for torch.norm

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81763
* #81762
* #81761
* __->__ #81765

This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

lezcano committed:
[PrimTorch] Add reference for torch.norm

This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -3146,6 +3146,33 @@ def equal(a: TensorLikeType, b: TensorLikeType) -> bool:
     return item(all(eq(a, b)))  # type: ignore[return-value]
 
 
+@register_decomposition(torch.ops.aten.norm)
```
cc @chillee, this seems to be using different overloads than original decomposition registration, is it ok?
Also, previous decomposition dispatched to ops that could be fused by nvfuser, whereas for this one it's unclear. Is this a concern and are there any tests we can run for this?

Chillee replied:
Yeah that's fine - this will just register it to all overloads (which is good now, since it also handles dtype).

But actually, I think this is the incorrect registration @Lezcano. I think `aten.norm` corresponds to `torch.linalg.vector_norm` :)

Lezcano replied:
You are completely right! 
Now, I assume that, since from the next PR in this stack we are going to have `norm` dispatching to `vector_norm` and `matrix_norm` and these are decomposed already, it should be fine not to register this as a decomposition, right?

Chillee replied:
Yeah this decomposition isn't needed for `__torch_dispatch__`.

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

Lezcano commented:
@ngimel this is ready for another review.

ngimel commented on torch/_decomp/decompositions.py:
```diff
@@ -1206,35 +1206,3 @@ def log_sigmoid_forward(self: Tensor) -> Tuple[Tensor, Tensor]:
-# The implementation matches torch.ops.aten.norm
-# torch.ops.aten.norm only supports numeric p, does not support Frobenius norm or nuclear norm
-# For 2-norm and -2 matrix norm, it doesn't compute the singular values, it just compute the norm the same as when p > 2.
-@register_decomposition([aten.norm.Scalar, aten.norm.ScalarOpt_dim])
```
we still need a decomposition for torch.ops.aten.norm, I don't think you can just delete it.

Lezcano replied:
I reimplemented it in terms of `vector_norm`. Later on in this stack I would like to make `at::linalg_vector_norm` the main implementation for this operation, and change all the calls from `at::norm` to calls into `at::linalg_vector_norm`. This will need moving the MPS and sparse implementations of `norm` to `linalg_vector_norm`, but I think we can try. It could serve as practice to see how much we can deprecate an implementation that's used left right and centre in practice.

Lezcano replied:
The diff following that one is a bit large, so here's an explanation:

After decomposing `norm` into `vector_norm`, I also needed to add a decomposition to `vector_norm`. I was going to add also a decomposition for `matrix_norm` and `svd`, but then I realised that I didn't need to do that given that they are ImplicitComposite so I added them to the relevant lists.

After doing that, I had to fix `upcast_tensor`, as it was not upcasting the `dtype` arg to `double` when it was `float`.
Then I started getting that the default implementation for the backward of `renorm` was more accurate than the reference implementation. I was able to fix it for CUDA touching a couple things in `renorm_backward`, but it's still failing on CPU. I ended up bumping a bit the tolerance for this operation, but while investigating this, I realised that the call to `torch.pow(x, 2)` within the ref of `vector_norm` was not being decomposed into a call to `mul(x,x)`, so I asked Ed to look into this. The mismatch on CPU may come from this? It'd be weird to me as `pow` does implement this optimisation on CPU, so I don't think so, but I don't know where is it coming from really... :/

ngimel replied:
This diff removes decomposition registration for `aten.norm` and this is not the state we want to be in. Each diff should not regress functionality.

Lezcano replied:
I added it back. There should be no regression

Lezcano replied:
Even more, I added support for the dtype arg.

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

Lezcano commented:
After trying to decompose `at::norm` into the more correct `linalg_vector_norm`, I started getting this error in  `test_proxy_tensor`:
```
  File "/opt/conda/lib/python3.7/site-packages/torch/_refs/linalg/__init__.py", line 78, in vector_norm
    if x.numel() == 0 and (ord < 0.0 or ord == float("inf")):
RuntimeError: Tensors of type TensorImpl do not have numel
```
I'm don't know how to fix it. CC @ezyang

ezyang commented:
This is probably related to @Chillee's most recent PR. I believe you should just xfail these.

@Chillee since most new ops are going to fail these tests we need to make it more obvious to people they should xfail them...

lezcano committed:
Update on "[PrimTorch] Add reference for torch.norm"


This ref does more things than `torch.norm`, and it fixes a few bugs
that `torch.norm` has. This implementation and the `torch.norm`
implementation come to terms in the next PR of this stack

We put this PR before, as otherwise `test_decomp.py` was failing.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge
